### PR TITLE
Button: Adopt support for border color, style, and width

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
## What?

Adopts border color, style, and width, supports for the Button block.

## Why?
- Provides greater control over button borders
- Allows any borders configured via theme.json to be tweaked via Global Styles.

## How?
- Opts into color, style, and width, support via the Button `block.json`

## Testing Instructions
1. Add some button styles via your theme.json `blocks` section. (See snippet below)
2. Load the editor add a `Buttons` block and a few buttons within it. They should reflect the styles in your theme.json
3. Save the post and switch to the site editor
4. Navigate to Global Styles > Blocks > Button > Layout and adjust the border there.
5. Switch back to your post, the buttons should now reflect the Global Styles values.
6. Select a button block and configure that individually via the inspector controls sidebar.
7. Ensure all the styles are applied correctly on the frontend.

Example theme.json snippet:
```json
	"styles": {
		"blocks": {
			"core/button": {
				"border": {
					"radius": "0",
					"color": "fuchsia",
					"style": "solid",
					"width": "10px"
				},
			}
		}
	}
```

## Screenshots or screencast 

https://user-images.githubusercontent.com/60436221/193013631-97c2d7bc-4b80-471c-8531-aaf3874f8e03.mp4

### CODE REVIEW

File: packages/block-library/src/button/editor.scss
Comment (aaronrobertshaw) [inline packages/block-library/src/button/editor.scss:83]: I'm not a huge fan of having to add this much CSS to provide defaults that can overcome styles in core. If anyone has any ideas on a better approach, I'd be happy to evolve this one.

Without these tweaks the user has to set multiple fields before a border shows up. Even now the default border color is `currentColor` so when the background is white and the button text is white, the border isn't really visible.

**Comment by glendaviesnz** [comment]
This is testing pretty well for me. Will have a think about alternatives to all those styles.

Do you think we need to worry about hover styles overriding the borders?

https://user-images.githubusercontent.com/3629020/193171656-edd58cd8-7e4f-4137-b295-026bea564c41.mp4

the above is on the Blockbase theme

**Comment by aaronrobertshaw** [comment]
> Do you think we need to worry about hover styles overriding the borders?

Good catch, thanks. I'd say we do.

Regarding theme.json and Global Styles, we might be able to leverage feature-level selectors to enforce the user's border selection. I don't think we have a means of styling the hover state of individual buttons at this point, so the block might be restricted to at least maintaining the same border.

We could extend the new styles further to enforce the border on hover when a border color, style or width are set. 

If we come up with a solution to more elegantly override the Element API provided button styles that might also provide a clue to handling pseudo selectors and hover/focus states set by theme.json.

**Comment by glendaviesnz** [comment]
> Without these tweaks the user has to set multiple fields before a border shows up

For me with all those styles removed it seems to work as expected as long as a border width is set, which makes sense to me, ie. I would not expect border color and border radius to be visible at all until a width is specified, ie. this flow makes sense to me:

https://user-images.githubusercontent.com/3629020/193175949-f2a7cd74-4f51-40e8-b82a-3a795e806a77.mp4

**Comment by aaronrobertshaw** [comment]
> ie. this flow makes sense to me:

I'll have to dig it up from the original border support PRs but this flow wasn't desired. Too many clicks and users expect to see visual feedback when they change a control.

**Comment by glendaviesnz** [comment]
> too many clicks and users expect to see visual feedback when they change a control.

In this instance this might be a `general` assumption ( which does generally make sense ) being applied to a `specific` context where it might not make sense, ie. a 0px border logically should not provide any visual feedback as it doesn't actually exist 😄

**Comment by aaronrobertshaw** [comment]
There's a lot of history around the border support and controls but here's a couple of links to where the application of a style and width when the user initially selects a color were raised. 

- https://github.com/WordPress/gutenberg/pull/31366#issuecomment-876983609
- https://github.com/WordPress/gutenberg/pull/31370#issuecomment-876973454

Another factor to consider here, if we remove the styles as suggested, is that this block now behaves differently from others with border support.

**Comment by glendaviesnz** [comment]
> Another factor to consider here, if we remove the styles as suggested, is that this block now behaves differently from others with border support.

yeh, if it has already been well discussed and decided on then this probably isn't the PR to be unwinding it all on

**Comment by aaronrobertshaw** [comment]
Thanks for the sanity check though. It is good to challenge past decisions and see how strongly they hold. I agree this might not be the PR to wind it back. 

I'll revisit this PR next week to see what we might be able to do about the pseudo-selector styles.

**Comment by glendaviesnz** [comment]
What is the status of this PR - is it ready for another test after you additional border supports?

**Comment by aaronrobertshaw** [comment]
> What is the status of this PR

Thanks for checking in on this one @glendaviesnz 👍 

I've rebased it and tested applying borders through theme.json or at the individual block level. It's all testing pretty well for me. If you could give it another once over, then I think we can land this pending your approval.

File: packages/block-library/src/button/editor.scss
Comment (glendaviesnz) [reply packages/block-library/src/button/editor.scss:83]: It doesn't seem like we need to replicate this in `editor.scss` and `styles.scss` - if I removed this section from here the CSS from `styles.scss` seemed to be applied in the editor and the frontend? I did a fresh build and a hard reload to make sure it wasn't just a case of the editor styles being cached.

**Comment by glendaviesnz** [comment]
It looks some of the core defaults (hover, focus) are still taking priority over the `theme.json` setting, but not the user applied settings:

https://user-images.githubusercontent.com/3629020/217147014-aba61d0e-4c86-4ca9-b29c-951e5f9790cf.mp4

this is with the theme.json settings suggested in the description. I can take a closer look at this tomorrow if you don't have chance to look at it.

**Comment by aaronrobertshaw** [comment]
Thanks for reviewing @glendaviesnz 👍 

> this is with the theme.json settings suggested in the description

I'm having trouble replicating the behaviour in your demo above. I'll continue to dig into it though. Using TwentyTwentyThree, here is what I see:


https://user-images.githubusercontent.com/60436221/217150621-2f73574a-e046-4772-8b8b-3ad5b8565640.mp4



In the meantime, it might also be worth noting that we can probably use the elements pseudo selectors within theme.json to configure the `:hover` and `:focus` border styles. Something like the following might work (will need to test this):

```json
	"core/button": {
		"border": {
			"radius": "0",
			"color": "fuchsia",
			"style": "solid",
			"width": "1em"
		},
		"elements": {
			"link": {
				":hover": {
					"border": {
						"color": "green"
					}
				}
			}
		}
	},
```

File: packages/block-library/src/button/editor.scss
Comment (aaronrobertshaw) [reply packages/block-library/src/button/editor.scss:83]: We do need to add this CSS to overcome [core's theme.json button element styles](https://github.com/WordPress/gutenberg/blob/d9da1250c77d3dc083729b72b8143dfc7bc222ba/lib/theme.json#L437) which reset `border-width` to `0`. 

The theme.json styles get the `.editor-styles-wrapper` class prefixed to their selectors in the editor. As a result, those styles override the styles in the frontend `style.scss`.

Without these styles, a user needs to adjust the border width before they can see any border. 

To witness this behaviour:
1. Delete or disable the styles here in the editor.scss file
2. Make sure you don't have any `core/button` styles in your theme.json
3. Edit a post and add a button
4. Select the button and select only a border color
5. Note that there is no visible border due to the core theme.json border width reset for button elements.



https://user-images.githubusercontent.com/60436221/217154597-f2d1ddb7-ed34-4cde-8cc5-2fdae382a5a3.mp4

**Comment by aaronrobertshaw** [comment]
> I can take a closer look at this tomorrow if you don't have chance to look at it.

@glendaviesnz If you can take a further look into the behaviour you were seeing that would be great thanks. 

I can't replicate any core styles overriding the button theme.json styles. At most, I noticed some themes that add their own styles with higher specificity than theme.json, overriding styles e.g. blockbase.

Apologies if I'm missing something obvious here 🤔

File: packages/block-library/src/button/editor.scss
Comment (glendaviesnz) [reply packages/block-library/src/button/editor.scss:83]: Thanks for the extra detail - I was adding the border width and then the color so wasn't picking this up - if I had read back through the old comments I would have picked all that up - sorry for wasting your time here! 

I have modified the comment to specify this detail like the one in style.scss does.

**Comment by glendaviesnz** [comment]
> @glendaviesnz If you can take a further look into the behaviour you were seeing that would be great thanks.

On further testing it was actually some higher specificity theme overrides causing this, which needs to be sorted at the theme end rather than in the block supports I think.

**Review by glendaviesnz — APPROVED** [review]
LGTM

**Pull request merged**
